proj/mnv2_first/gateware/store.py

```diff
@@ -0,0 +1,186 @@
+        return statement_list
+
+    def elab(self, m):
+        m.d.comb += [
+            self.done.eq(True),
+            self.w_addr.eq(self.count[self.num_memories_bits:])
+        ]
+        with m.If(self.restart):
+            m.d.sync += self.count.eq(0)
+        with m.Elif(self.start):
+            m.d.comb += [
+                self.w_en[self.count[:self.num_memories_bits]].eq(1),
+                self.w_data.eq(self.in0),
+            ]
+            m.d.sync += self.count.eq(self.count + 1)
+
+
+class CircularIncrementer(SimpleElaboratable):
+    """Does circular increments of a memory address counter over a store.
+
+    Parameters
+    ----------
+    depth:
+        Maximum number of items stored in memory
+
+    Public Interface
+    ----------------
+    restart: Signal input
+        Signal to reset address to zero
+    next: Signal input
+        Produce next piece of data (available next cycle).
+    limit: Signal(range(depth))) input
+        Number of items stored in memory
+    r_addr: Signal(range(depth)) output
+        Next address
+    """
+
+    def __init__(self, depth):
+        self.depth = depth
+        self.restart = Signal()
+        self.next = Signal()
+        self.limit = Signal(range(depth))
+        self.r_addr = Signal(range(depth))
+
+    def elab(self, m):
+        # Current r_addr is the address being presented to memory this cycle
+        # By default current address is last address, but this can be
+        # modied by the reatart or next signals
+        last_addr = Signal.like(self.r_addr)
+        m.d.sync += last_addr.eq(self.r_addr)
+        m.d.comb += self.r_addr.eq(last_addr)
+
+        # Respond to inputs
+        with m.If(self.restart):
+            m.d.comb += self.r_addr.eq(0)
+        with m.Elif(self.next):
+            m.d.comb += self.r_addr.eq(Mux(last_addr >=
+                                           self.limit - 1, 0, last_addr + 1))
+
+
+class FilterValueGetter(Xetter):
+    """Gets next word from 4-way filter value store.
+
+    This is a temporary getter. The full accelerator will use filter values internally.
+
+    Parameters:
+    num_memories:
+        number of memories in the store. data will be striped over the memories.
+        Assumed to be a power of two
+    depth:
+        Maximum number of items stored in each memory
+
+    Attributes:
+        r_addr: Signal(range(depth/num_memories)) output
+            Current read pointer
+        r_bank: Signal(range(num_memories)) output
+            Which of the memories to get from
+        restart: Signal() input
+    """
+
+    def __init__(self, num_memories, depth):
+        super().__init__()
+        assert 0 == ((num_memories - 1) &
+                     num_memories), "Num memories (f{num_memories}) not a power of two"
+        self.num_memories = num_memories
+        self.r_addr = Signal(range(depth))
+        self.r_bank = Signal(range(num_memories))
+        self.limit = Signal(self.num_memories * depth)
```

Comment:
Here is the issue.  `self.limit` is used below to determine with width of `count`.